crates/autopilot/src/domain/settlement/surplus.rs

```diff
@@ -0,0 +1,120 @@
+//! Settlement surplus calculation
+
+use {
+    super::Encoded,
+    crate::domain::{auction, auction::order, eth},
+    num::BigRational,
+    number::conversions::big_rational_to_u256,
+    shared::conversions::U256Ext,
+    std::collections::HashMap,
+};
+
+/// Settlement surplus
+///
+/// Denominated in surplus tokens. Contains multiple values since settlement can
+/// have multiple orders with different tokens.
+pub struct Surplus {
+    surplus: HashMap<eth::TokenAddress, eth::TokenAmount>,
+}
+
+impl Surplus {
+    pub fn new(settlement: &Encoded) -> Self {
+        let tokens = settlement.tokens();
+        let clearing_prices = settlement.clearing_prices();
+
+        let mut surplus: HashMap<eth::TokenAddress, eth::TokenAmount> = Default::default();
+        for trade in settlement.trades() {
+            let clearing_prices = ClearingPrices {
+                sell: clearing_prices[trade.sell_token_index],
+                buy: clearing_prices[trade.buy_token_index],
+            };
+
+            let trade_surplus = trade_surplus(
+                trade.flags.order_kind(),
+                trade.executed,
+                trade.sell_amount,
+                trade.buy_amount,
+                &clearing_prices,
+            )
+            .map(|surplus| match trade.flags.order_kind() {
+                order::Kind::Buy => eth::Asset {
+                    amount: surplus.into(),
+                    token: tokens[trade.sell_token_index],
+                },
+                order::Kind::Sell => eth::Asset {
+                    amount: surplus.into(),
+                    token: tokens[trade.buy_token_index],
+                },
+            });
+
+            match trade_surplus {
+                Some(trade_surplus) => {
+                    *surplus.entry(trade_surplus.token).or_default() += trade_surplus.amount;
+                }
+                None => tracing::warn!("surplus failed for trade {:?}", trade.order_uid),
+            };
+        }
+
+        Self { surplus }
+    }
+
+    /// Surplus denominated in the native token (ETH)
+    pub fn normalized_with(
```

Comment:
I think this method would make more sense on the `Settlement` struct directly. This way we don't have to make the non-normalized surplus public and reduce the number of concepts consumers of the settlement domain object have to work with.

Reply:
Makes sense. Currently, we indeed make no use of the non-normalized surplus.

Reply:
done